File: backend/dataset/viewer/channels.py
"""Channel and camera-stream structure, derived from `meta/info.json` alone.

`FR-DAT-011` acceptance ①: stream count and type are read from the dataset's
`info.json`, never assumed from a fixed slot list. This module is that derivation
— it enumerates the `observation.images.*` features, classifies each as RGB or
depth, and parses the `observation.state`/`action` channel names into per-motor,
unit-labelled channels. The camera identifier and image-key grammar are
`CTR-PRIM@v1` primitives, imported from `contracts.prim`; the unit convention is
mirrored from `CTR-REC@v1` through the `names` strings (see `constants`).

`FR-DAT-012`/`FR-DAT-016` live here too: axis units come from the channel-name
suffix, and following error is defined on position channels only — a `.vel`/
`.torque` name appearing in `action` is the FAIL_BLOCKING poison and is refused,
never plotted as a "leader-measured torque".
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Any

from backend.dataset.viewer.constants import (
    ACTION_KEY,
    FEATURE_IS_DEPTH_KEY,
    FEATURE_NAMES_KEY,
    FEATURE_SHAPE_KEY,
    OBSERVATION_STATE_KEY,
    POSITION_SUFFIX,
    SUFFIX_UNITS,
    TORQUE_SUFFIX,
    UNKNOWN_UNIT,
    VELOCITY_SUFFIX,
)
from contracts.prim import (
    DEPTH_KEY_SUFFIX,
    IMAGE_KEY_PREFIX,
    CameraSlotKey,
    FrameType,
    slot_from_image_key,
)


class ViewerChannelError(ValueError):
    """Raised when `info.json` channel structure is unreadable or self-contradictory.

    Distinct from a layout/IO error: this is a malformed feature set (a missing
    `names` list, or an `action` feature carrying a torque channel), not a missing
    file. An `action.torque` name is the `07` §2.3.3 FAIL_BLOCKING case surfaced
    at read time rather than silently plotted.
    """


def unit_for_channel(name: str) -> str:
    """Return the display unit for a channel, from its name suffix.

    Args:
        name: A channel name such as `left_joint_1.pos` or `right_gripper.torque`.

    Returns:
        (str) `deg`/`deg/s`/`Nm` per the `CTR-REC@v1` convention, or `UNKNOWN_UNIT`
            when the name carries no recognised suffix (shown, never blanked).
    """
    for suffix, unit in SUFFIX_UNITS.items():
        if name.endswith(suffix):
            return unit
    return UNKNOWN_UNIT


def axis_label(name: str) -> str:
    """Return a channel's axis label with its unit, e.g. `left_joint_1.pos [deg]`.

    Args:
        name: A channel name.

    Returns:
        (str) The name annotated with its bracketed unit (`FR-DAT-016`).
    """
    return f"{name} [{unit_for_channel(name)}]"


@dataclass(frozen=True)
class CameraStream:
    """One configured camera stream, as `info.json` declares it.

    Attributes:
        image_key: The full `observation.images.<slot>[_depth]` feature key.
        slot: The base camera slot the stream belongs to (RGB and depth of one
            camera share a slot), a `CTR-PRIM@v1` `CameraSlotKey`.
        frame_type: RGB or DEPTH, the `CTR-PRIM@v1` frame-type primitive.
        shape: The `[height, width, channels]` shape from the feature body.
    """

    image_key: str
    slot: CameraSlotKey
    frame_type: FrameType
    shape: tuple[int | str, ...]

    @property
    def is_depth(self) -> bool:
        """Whether this stream carries depth rather than RGB."""
        return self.frame_type == FrameType.DEPTH


@dataclass(frozen=True)
class FollowingErrorPair:
    """A position channel's follower/leader index pair for following error.

    Following error is `observation.state[<motor>.pos] - action[<motor>.pos]` and
    is defined on position channels only (`FR-DAT-012`); torque and velocity have
    no `action` counterpart and never enter this pairing.

    Attributes:
        motor: The motor key the pair belongs to (e.g. `left_gripper`).
        state_index: Column of `<motor>.pos` in `observation.state`.
        action_index: Column of `<motor>.pos` in `action`.
    """

    motor: str
    state_index: int
    action_index: int


def _feature_body(features: Mapping[str, Any], key: str) -> Mapping[str, Any]:
    """Return a feature body, or raise when it is absent or not a mapping."""
    body = features.get(key)
    if not isinstance(body, Mapping):
        raise ViewerChannelError(f"info.json feature {key!r} is missing or not an object")
    return body


def _names(features: Mapping[str, Any], key: str) -> tuple[str, ...]:
    """Return a feature's channel `names`, or raise when absent."""
    body = _feature_body(features, key)
    names = body.get(FEATURE_NAMES_KEY)
    if not isinstance(names, Sequence) or isinstance(names, str) or not names:
        raise ViewerChannelError(
            f"info.json feature {key!r} has no usable {FEATURE_NAMES_KEY!r} list"
        )
    return tuple(str(name) for name in names)


def state_channel_names(features: Mapping[str, Any]) -> tuple[str, ...]:
    """Return the `observation.state` channel names in dataset order."""
    return _names(features, OBSERVATION_STATE_KEY)


def action_channel_names(features: Mapping[str, Any]) -> tuple[str, ...]:
    """Return the `action` channel names, verifying they are position only.

    Raises:
        ViewerChannelError: If any `action` name carries a `.vel`/`.torque`
            suffix — the `07` §2.3.3 poison the viewer must never display as a
            leader-measured value.
    """
    names = _names(features, ACTION_KEY)
    poisoned = [name for name in names if name.endswith((VELOCITY_SUFFIX, TORQUE_SUFFIX))]
    if poisoned:
        raise ViewerChannelError(
            f"action carries non-position channels {poisoned}; action is position only "
            "and a .vel/.torque value is neither a robot command nor a leader-measured torque"
        )
    return names


def following_error_pairs(features: Mapping[str, Any]) -> tuple[FollowingErrorPair, ...]:
    """Pair each position channel across `observation.state` and `action`.

    Only `.pos` channels are paired; a motor whose `.pos` appears in one feature
    but not the other is skipped rather than guessed. The result drives per-joint
    following error (`FR-DAT-012`).

    Args:
        features: The `info.json` feature map.

    Returns:
        (tuple[FollowingErrorPair, ...]) One pair per position channel shared by
            both features, in `observation.state` order.
    """
    state_names = state_channel_names(features)
    action_names = action_channel_names(features)
    action_index = {name: index for index, name in enumerate(action_names)}

    pairs: list[FollowingErrorPair] = []
    for state_idx, name in enumerate(state_names):
        if not name.endswith(POSITION_SUFFIX):
            continue
        act_idx = action_index.get(name)
        if act_idx is None:
            continue
        motor = name[: -len(POSITION_SUFFIX)]
        pairs.append(FollowingErrorPair(motor=motor, state_index=state_idx, action_index=act_idx))
    return tuple(pairs)


def _classify_frame_type(image_key: str, body: Mapping[str, Any]) -> FrameType:
    """Classify a stream as depth or RGB from `is_depth_map`, then the key suffix.

    The `is_depth_map` flag is authoritative; the `_depth` key suffix is the
    fallback for a dataset that omitted the flag. Both agree by construction under
    `CTR-REC@v1`, so this only diverges on a malformed dataset, where the flag wins.
    """
    flagged = bool(body.get(FEATURE_IS_DEPTH_KEY, False))
    if flagged or image_key.endswith(DEPTH_KEY_SUFFIX):
        return FrameType.DEPTH
    return FrameType.RGB


def _base_slot(image_key: str, frame_type: FrameType) -> CameraSlotKey:
    """Recover the base camera slot from an image key, stripping any depth suffix.

    RGB and depth of one camera share a slot; the depth key carries the `_depth`
    suffix on top of the slot, so it is stripped before the `CTR-PRIM@v1` grammar
    recovers the slot.
    """
    if frame_type == FrameType.DEPTH and image_key.endswith(DEPTH_KEY_SUFFIX):
        return CameraSlotKey(image_key[len(IMAGE_KEY_PREFIX) : -len(DEPTH_KEY_SUFFIX)])
    return slot_from_image_key(image_key)


def camera_streams(features: Mapping[str, Any]) -> tuple[CameraStream, ...]:
    """Enumerate the configured camera streams from `info.json`.

    A stream is any feature whose key carries the `CTR-PRIM@v1` image prefix. The
    count and RGB/depth split come from the feature set, never a fixed slot table
    (`FR-DAT-011` acceptance ①). Streams are returned in feature-declaration order.

    Args:
        features: The `info.json` feature map.

    Returns:
        (tuple[CameraStream, ...]) The configured streams, RGB and depth alike.
    """
    streams: list[CameraStream] = []
    for image_key, body in features.items():
        if not image_key.startswith(IMAGE_KEY_PREFIX):
            continue
        if not isinstance(body, Mapping):
            raise ViewerChannelError(f"image feature {image_key!r} body is not an object")
        frame_type = _classify_frame_type(image_key, body)
        raw_shape = body.get(FEATURE_SHAPE_KEY, [])
        # info.json shapes carry symbolic dims ("height"/"width") beside integer
        # channel counts; both are preserved as declared, not coerced away.
        shape: tuple[int | str, ...] = (
            tuple(dim if isinstance(dim, str) else int(dim) for dim in raw_shape)
            if isinstance(raw_shape, Sequence) and not isinstance(raw_shape, str)
            else ()
        )
        streams.append(
            CameraStream(
                image_key=image_key,
                slot=_base_slot(image_key, frame_type),
                frame_type=frame_type,
                shape=shape,
            )
        )
    return tuple(streams)

File: frontend/src/screens/S-08/channels.ts
// The names-indexed channel resolution — the load-bearing rule of S-08 (CG-G-S08a/g).
//
// observation.state is ONE vector whose columns mix three units. Which column is
// `left_joint_1.pos` and which is `right_gripper.torque` is declared by the
// info.json `names` list, and it MOVES the instant `use_velocity_and_torque` toggles
// the state between its 24-dim (pos+vel+torque) and 8-dim (pos-only) shape. So a
// channel is only ever resolved by `stateNames.indexOf(name)` — a compiled-in column
// index would silently scramble every plot on that toggle (CG-G-S08a).
//
// The unit convention (.pos->deg, .vel->deg/s, .torque->Nm) is CTR-REC@v1's, consumed
// through the `names` suffixes exactly as the backend viewer does (backend/dataset/
// viewer/constants.py). `contract.test.ts` reads that backend constants file and
// asserts this mirror equals it, so a convention change fails the lane rather than
// drifting. `observation.effort` is deliberately absent — it is not a key that exists
// (CG-G-S08b), so nothing here names it.

import type { EpisodeSignals } from "./types";

// The per-motor channel suffixes and their display units (CTR-REC@v1, mirrored via
// the info.json name suffixes). A single state vector mixes all three, so an
// unlabelled plot misreads a torque as degrees — every axis label carries its unit.
export const POSITION_SUFFIX = ".pos";
export const VELOCITY_SUFFIX = ".vel";
export const TORQUE_SUFFIX = ".torque";

export const POSITION_UNIT = "deg";
export const VELOCITY_UNIT = "deg/s";
export const TORQUE_UNIT = "Nm";

// Ordered longest-suffix-first is unnecessary here (the three suffixes share no
// prefix), but the map is the single source both `unitForChannel` and the contract
// mirror read, so the convention lives in exactly one place.
export const SUFFIX_UNITS: ReadonlyArray<readonly [string, string]> = [
  [POSITION_SUFFIX, POSITION_UNIT],
  [VELOCITY_SUFFIX, VELOCITY_UNIT],
  [TORQUE_SUFFIX, TORQUE_UNIT],
];

// Shown when a channel name carries no recognised suffix — never silently blank, so a
// missing unit is visible rather than mistaken for dimensionless (backend UNKNOWN_UNIT).
export const UNKNOWN_UNIT = "?";

// Return a channel's display unit from its name suffix.
export function unitForChannel(name: string): string {
  for (const [suffix, unit] of SUFFIX_UNITS) {
    if (name.endsWith(suffix)) {
      return unit;
    }
  }
  return UNKNOWN_UNIT;
}

// Return a channel's axis label with its bracketed unit, e.g. `left_joint_1.pos [deg]`
// — the per-channel unit annotation a mixed-unit state vector needs (CG-G-S08g).
export function axisLabel(name: string): string {
  return `${name} [${unitForChannel(name)}]`;
}

// Raised when a channel name is asked for that the dataset's `names` list does not
// carry — surfaced rather than returning a wrong column (which is exactly what a fixed
// index would do after the dimension toggles).
export class ChannelResolutionError extends Error {
  readonly channel: string;

  constructor(channel: string) {
    super(`'${channel}' is not an observation.state channel of this dataset`);
    this.name = "ChannelResolutionError";
    this.channel = channel;
  }
}

// One resolved channel series: the per-frame values pulled by the `names` index, plus
// the channel's unit for the axis label.
export interface ChannelSeries {
  name: string;
  values: readonly number[];
  unit: string;
}

// Resolve one observation.state channel's series by its `names` index — the only
// permitted access (CG-G-S08a). The column is `signals.stateNames.indexOf(name)`,
// computed per dataset, so the same channel name lands on the right column whether the
// dataset is 8-dim or 24-dim.
export function resolveStateChannel(signals: EpisodeSignals, name: string): ChannelSeries {
  const index = signals.stateNames.indexOf(name);
  if (index < 0) {
    throw new ChannelResolutionError(name);
  }
  const values = signals.state.map((frame) => frame[index]);
  return { name, values, unit: unitForChannel(name) };
}
